### PR TITLE
Fix media uploads for existing subjects

### LINKED ISSUE #239: Problem uploading new media to subjects with existing media
**Problem:** When calling `Subject.add_location()` for a subject that has already been created and saved (i.e., has ID + existing media `locations` entries), new media is not uploaded correctly.

**Current Behavior:** When initialized (e.g., using `Subject.find()`), subjects with existing `locations` entries (N_locations > 0) are initialized with an empty `_media_files` list (N_mediafiles = 0).  When `Subject.save()` is executed after adding a new subject image via `add_location()`, the uninitialized `_media_files` list doesn't match up element-for-element with the `locations` array leading to issues in the `zip()` call below:
https://github.com/zooniverse/panoptes-python-client/blob/c0bc5c2f07b1cdcbbed21b76e39efab8cd406019/panoptes_client/subject.py#L158-L161

Furthermore, `locations` is not added to `modified_attributes`.  But even if `modified_attributes` is altered manually, the `save()` will fail when the mismatch leads to the client trying to PUT to an existing, unwritable location.

Here's an illustrative example:
```
s = Subject.find(<SUBJECT_ID>)                                              

In [7]: len(s.locations)                                                        
Out[7]: 3
In [8]: len(s._media_files)                                                     
Out[8]: 0

s.add_location('kitten.jpg')                                            

In [10]: len(s.locations)                                                       
Out[10]: 4
In [11]: len(s._media_files)                                                    
Out[11]: 1
```

**Desired Behavior:** Initialize `Subject()` instances with a `_media_files` array whose length matches that of `locations` array and is filled with `None` elements.  Also, add `locations` to `modified_attributes` when `add_location` is called.  Fixing both of these issues solves upload problem.

### PR DESCRIPTION
**Pull request by lcjohnso**
Closes #239

Initialize existing `Subject()` instances with a `_media_files` array whose length matches that of `locations` array and is filled with `None` elements. Also, add `locations` to `modified_attributes` when `add_location` is called. These two edits fixes issue where `add_location()` calls for existing subjects -- those that already have been saved with existing media `locations` -- was failing.

### CODE REVIEW

File: panoptes_client/subject.py [hunk 1]
```diff
@@ -104,6 +104,8 @@ def __init__(self, raw={}, etag=None):
             self.metadata = {}
             self._original_metadata = {}
         self._media_files = []
+        for _l in self.locations:
+            self._media_files.append(None)
```
Comment (adammcmaster) [inline panoptes_client/subject.py:108]:
```suggestion
        self._media_files = [None] * len(self.locations)
```

**Comment by lcjohnso** [comment]
Coming back to this after doing a bit more hacking on location updates.

Adam's suggestion for a fix works great, but I'd like to do a little more testing to better understand when rows in the Panoptes `media` table are edited versus rewritten.  I'm worried that the current setup may orphan `media` entries when the URL in the media dict doesn't exactly match the one in the table.  Specifically, whether or not the URL has the `https://panoptes-uploads.zooniverse.org/` prefix.

File: panoptes_client/subject.py [hunk 2]
```diff
@@ -175,6 +175,9 @@ def save(self, client=None):
                             ),
                             log_args=False,
                         )
+
+                self._media_files = [None] * len(self.locations)
+
```
[the remark this answers is not in the document]
Comment (lcjohnso) [reply panoptes_client/subject.py:180]: ok ok

**Comment by lcjohnso** [comment]
After rebasing and making change suggested by Adam above (to allow for repeated `locations` updates via reinitializing `_media_files`), I confirmed my above suspicions: existing media locations are treated as **external media** after the Subject update, due to the fact that their `locations` entry is a hash and is processed by [this code in the subject model](https://github.com/zooniverse/panoptes/blob/e3949292f9510224cb3688bb362250bed7967d3a/app/models/subject.rb#L42-L47).

There are two outcomes of this that I wanted to highlight:
1. The new "external" media entries point to Azure media items using their full valid URLs (e.g., https://panoptes-uploads.zooniverse.org/subject_location/752360a5-455e-4f3e-9e3f-0ec69d10b764.png), which are `media_url`s returned by the Azure adapter's `get_path()` function ([link to code](https://github.com/zooniverse/panoptes/blob/6ef1579c7f9030cd76b8b01722744670bce4f338/app/services/media_storage/azure_adapter.rb#L34)).
2. When existing subjects are updated via a `PUT` request as part of the subject `save()`, all of the old `locations` are orphaned in the `media` DB table (`linked_id` set to `NULL`).  There are 459k rows in the `media` table that have this state (maybe not great?). But as an upside: the media files themselves are not deleted, which is good because the hosted files are now being pointed to by the new "external media" entries.

In conclusion: if we are OK with the current orphan + external link consequences, this feature is ready to go!

**Comment by PmasonFF** [comment]
I have tried to build a modified panoptes client with this changed subject.py replacing the original in a venv to test this out. Using this code I can add images - **sometimes**.

 So first generate a subject:  This code works -
````
""" create new subjects - useful for testing media"""
subject = Subject()
subject.links.project = project
subject.locations = \
    [{'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/daaa5cb4-3932-459e-ab1e-d5d97a58ef16.jpeg'},
{'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/7341ae09-3fa3-4933-9368-1e272d8775a6.jpeg'},
{'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/1100dbf6-52ef-4c82-b049-4afc232c4e8d.jpeg'}]
subject.save()
subject_revised = Subject(subject.id)
print(subject_revised.id, subject_revised.locations)
````
Assume it created subject 75224285

Then this code works fine: (As recently used to fix a subject set for the Racial Restrictive Covenants project back in 03-03-2022)  This code works for this subject in the release panoptes-client as well
````
""" remove a image from a multiframe subject"""
subject = Subject(75224285)
print(subject.id, subject.locations)
subject.locations = subject.locations[:-1]
print(subject.id, subject.locations)
subject.save()
subject_revised = Subject(75224285)
print(subject_revised.id, subject_revised.locations)
````

Then using this revised subject,py I can get this to work though it does not work in the release version since the server change:
````
""" add a image to a multiframe subject"""
subject = Subject(75224285)
print(subject.id, subject.locations)
subject.add_location({'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/1100dbf6-52ef-4c82-b049-4afc232c4e8d.jpeg'})
print(subject.id, subject.locations)  # - so far so good!
subject.save()
subject_revised = Subject(75224285)
print(subject_revised.id, subject_revised.locations)
````
### I can remove and add images to this subject at will.  BUT for other subjects such as 75076670, I get this error when I try to remove a image:
````
Traceback (most recent call last):
  File "C:/py_scripts/Scripts_Panoptes_utilities/work_with_subjects.py", line 213, in <module>
    subject.save()
  File "C:\Users\User1\AppData\Roaming\Python\Python37\site-packages\panoptes_client\subject.py", line 145, in save
    log_args=False,
  File "C:\Users\User1\AppData\Local\Programs\Python\Python37-32\lib\site-packages\redo\__init__.py", line 170, in retry
    return action(*args, **kwargs)
  File "C:\Users\User1\AppData\Roaming\Python\Python37\site-packages\panoptes_client\panoptes.py", line 817, in save
    etag=self.etag
  File "C:\Users\User1\AppData\Roaming\Python\Python37\site-packages\panoptes_client\panoptes.py", line 364, in put
    retry=retry,
  File "C:\Users\User1\AppData\Roaming\Python\Python37\site-packages\panoptes_client\panoptes.py", line 269, in json_request
    retry=retry,
  File "C:\Users\User1\AppData\Roaming\Python\Python37\site-packages\panoptes_client\panoptes.py", line 245, in http_request
    response.status_code
panoptes_client.panoptes.PanoptesAPIException: Received HTTP status code 500 from API
````

When I try to add an image to this subject I get the exact same error message.  Testing in an unadulterated panoptes client I get the same error on attempting to remove an image from 75076670, but when I try to add an image there,  it simply does not work and there is no error message - as expected since this problem first arose. 

### So bottom line this fix does not work with all subjects.  

Here are some more that it does not work with: 74820158, 74819168, 75075504.

I have a bunch of subjects I need to fix for nebraska-wildlife-watch and it happens they are the "don't work" type.
Thanks
Peter

**Comment by yuenmichelle1** [comment]
I believe the 500 error is an API issue due to changes made ~20 days ago to ensure `priority` key in metadata is downcased. Issue linked here: https://github.com/zooniverse/panoptes/issues/3832. The PUT request from the panoptes-python-client is assuming that metadata will also be changed (i.e. `metadata` would also be params within the PUT request). A fix for linked issue should be coming to Panoptes soon.

**Comment by PmasonFF** [comment]
Yeah! This work-around does it!
````
subject = Subject(74819168)
print(subject.id, subject.locations, subject.metadata)
metadata = subject.metadata
subject.locations = \
    [{'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/daaa5cb4-3932-459e-ab1e-d5d97a58ef16.jpeg'},
     {'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/7341ae09-3fa3-4933-9368-1e272d8775a6.jpeg'},
     {'image/jpeg': 'https://panoptes-uploads.zooniverse.org/subject_location/1100dbf6-52ef-4c82-b049-4afc232c4e8d.jpeg'}
     ]
subject.metadata = metadata
print(subject.id, subject.locations, subject.metadata)  # - so far so good!
subject.save()
subject_revised = Subject(74819168)
print(subject_revised.id, subject_revised.locations, subject_revised.metadata)
````
Did not seem to work with just a subject.metadata.update(metadata)

  Thanks @yuenmichelle1

**Review by zwolf — APPROVED** [review]

**Comment by camallen** [comment]
I've merged and deployed https://github.com/zooniverse/panoptes/pull/3834 so the work around shouldn't be needed anymore.
